File: 0202_Problem_1.hs
-- code_report Solution
-- Problem Link (Contest):  https://leetcode.com/contest/weekly-contest-202/problems/three-consecutive-odds/
-- Problem Link (Practice): https://leetcode.com/problems/three-consecutive-odds/

-- Note this problem is very similar to MCO (Max Consecutive Ones)

import Data.List (groupBy)
import Data.Function (on)

threeConsecutiveOdds :: [Int] -> Bool
threeConsecutiveOdds = (3<=) 
                     . maximum 
                     . map length 
                     . filter (odd . head)   -- (.) is the B (Bluebird) combinator
                     . groupBy (on (==) odd) -- on  is the P (Psi)      combinator

File: 0202_Problem_1.clj
;; code_report Solution
;; Problem Link (Contest):  https://leetcode.com/contest/weekly-contest-202/problems/three-consecutive-odds/
;; Problem Link (Practice): https://leetcode.com/problems/three-consecutive-odds/

;; Note this problem is very similar to MCO (Max Consecutive Ones)

(defn three-consecutive-odds [arr]
  (->> arr
       (partition-by odd?)
       (filter #(odd? (first %)))
       (map count)
       (apply max)
       (<= 3)))
